include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h

```diff
@@ -36,24 +36,59 @@ class sycl_backend
     using resource_container_t = std::vector<execution_resource_t>;
 
   private:
-    class async_waiter
+    class async_waiter_base{
+        public:
+            virtual void wait() = 0;
+            virtual void report() = 0;
+            virtual bool is_complete() = 0;
+    };
+
+    std::vector<async_waiter_base*> async_waiter_arr;
+
+    template<typename T>
+    void add_waiter(T *t){
```

Comment:
Are you protecting this?  Is it thread safe?

Reply:
Fixed